File: content/snippets/go/design_patterns_strategy.go
package main

import "fmt"

type Strategy interface {
	Execute(int, int) int
}

type AddStrategy struct{}

func (s *AddStrategy) Execute(a, b int) int {
	return a + b
}

type SubtractStrategy struct{}

func (s *SubtractStrategy) Execute(a, b int) int {
	return a - b
}

type MultiplyStrategy struct{}

func (s *MultiplyStrategy) Execute(a, b int) int {
	return a * b
}

type Context struct {
	strategy Strategy
}

func (c *Context) SetStrategy(s Strategy) {
	c.strategy = s
}

func (c *Context) ExecuteStrategy(a, b int) int {
	return c.strategy.Execute(a, b)
}

func main() {
	ctx := &Context{}

	ctx.SetStrategy(&AddStrategy{})
	fmt.Printf("Add: %d\n", ctx.ExecuteStrategy(5, 3))

	ctx.SetStrategy(&SubtractStrategy{})
	fmt.Printf("Subtract: %d\n", ctx.ExecuteStrategy(5, 3))

	ctx.SetStrategy(&MultiplyStrategy{})
	fmt.Printf("Multiply: %d\n", ctx.ExecuteStrategy(5, 3))
}

File: content/snippets/typescript/design_patterns_strategy.ts
interface IStrategy {
    execute(a: number, b: number): number;
}

class AddStrategy implements IStrategy {
    execute(a: number, b: number): number {
        return a + b;
    }
}

class SubtractStrategy implements IStrategy {
    execute(a: number, b: number): number {
        return a - b;
    }
}

class MultiplyStrategy implements IStrategy {
    execute(a: number, b: number): number {
        return a * b;
    }
}

class Context {
    private strategy: IStrategy;
    
    setStrategy(strategy: IStrategy): void {
        this.strategy = strategy;
    }
    
    executeStrategy(a: number, b: number): number {
        return this.strategy.execute(a, b);
    }
}

const ctx = new Context();

ctx.setStrategy(new AddStrategy());
console.log(`Add: ${ctx.executeStrategy(5, 3)}`);

ctx.setStrategy(new SubtractStrategy());
console.log(`Subtract: ${ctx.executeStrategy(5, 3)}`);

ctx.setStrategy(new MultiplyStrategy());
console.log(`Multiply: ${ctx.executeStrategy(5, 3)}`);
